gpii/node_modules/preferencesServer/src/preferencesService.js

```diff
@@ -0,0 +1,359 @@
+     * Create a gpiiKey record with the provided input.
+     * @param addGpiiKey {Function} The addGpiiKey() API provided by gpii.oauth2.dbDataStore
+     * @param input {Object} The data passed on from the previous step. Its structure:
+     *  {
+     *      gpiiKey: {String},
+     *      prefsSafeId: {String},
+     *      prefsSafeDataAdded: {
+     *          prefsSafeType: {String}
+     *          name: {String}
+     *          password: {String}
+     *          email: {String}
+     *          preferences: {Object}  // must be provided
+     *      }
+     *  }
+     * @return {Promise} A promise object whose resolved value is:
+     *  {
+     *      gpiiKey: {String},
+     *      preferences: {Object}
+     *  }
+     */
+    gpii.preferencesServer.preferencesService.createGpiiKey = function (addGpiiKey, input) {
+        var addGpiiKeyPromise = addGpiiKey({
+            gpiiKey: input.gpiiKey,
+            prefsSafeId: input.prefsSafeId,
+            prefsSetId: gpii.preferencesServer.defaultprefsSetId
+        });
+
+        var mapper = function (addGpiiKeyResponse) {
+            fluid.log("Preferences Service, created a new GPII key with id: ", addGpiiKeyResponse.id);
+            return {
+                gpiiKey: addGpiiKeyResponse.id,
+                preferences: input.prefsSafeDataAdded.preferences
+            };
+        };
+        return fluid.promise.map(addGpiiKeyPromise, mapper);
+    };
+
+    /**
+     * The entry point of updatePreferences() API.
+     * It handles:
+     * 1. If an existing prefs safe already associates with the GPII key, update the prefs safe;
+     * 2. If no prefs safe is associated with the GPII key, create the prefs safe and associate with the GPII key.
+     * @param that {Component} An instance of gpii.preferencesServer.preferencesService
+     * @param preferences {Object} The preferences to be updated
+     * @param gpiiKey {String} Optional. The new GPII key, an UUID will be generated if it's not provided
+     * @param mergePrefs {Boolean} A flag that indicates whether to merge the incoming preferences with the existing preferences
+     * @return {Promise} A promise object whose resolved value is:
+     *  {
+     *      gpiiKey: {String},
+     *      preferences: {Object}
+     *  }
+     */
+    gpii.preferencesServer.preferencesService.updatePreferences = function (that, preferences, gpiiKey, mergePrefs) {
+        mergePrefs = mergePrefs || false;  // By default, do not merge with the incoming preferences with the existing preferences
+        var promiseTogo = fluid.promise();
+        var gpiiKeyPromise = that.dataStore.findPrefsSafeByGpiiKey(gpiiKey);
+
+        gpiiKeyPromise.then(function (data) {
+            var updatePromise = fluid.promise();
+            var prefsSafeId = fluid.get(data, ["gpiiKeyDetails", "prefsSafeId"]);
+
+            if (prefsSafeId) {
+                if (data.prefsSafe.prefsSafeType === gpii.preferencesServer.prefsSafeType.snapset) {
```

Comment:
I think I would invert this logic and test that we can only modify `prefsSafeType === "user"` (white-list vs black-list).

That way, in the future, if we add other `prefsSafeTypes`, we default to read-only and have to consciously enable modification.